### PR TITLE
fix(controller): use update instead of patch

### PR DESCRIPTION
**Pull request by cndoit18**
Signed-off-by: cndoit18 <cndoit18@outlook.com>

Fixes: #825
Fixes: #824
Fixes: #743

### CODE REVIEW

File: pkg/controller/apply.go [hunk 1]
```diff
@@ -47,8 +45,7 @@ func (c *Controller) Apply(ctx context.Context, res *Resource) error {
 
 	l.Info("apply changing", "key", key, "kind", resource.GetObjectKind().GroupVersionKind())
 
-	if err := c.Client.Patch(ctx, resource, client.Apply, &client.PatchOptions{
-		Force:        &force,
+	if err := c.Client.Update(ctx, resource, &client.UpdateOptions{
```
Comment (steven-zou) [inline pkg/controller/apply.go:48]: So switch back to update and then append a new mutation to fix the problems caused by using update? @cndoit18

Comment (cndoit18) [reply pkg/controller/apply.go:48]: > 

right

**Comment by bitsf** [comment]
we need to do more test this first, the fix is a little tricky.

File: pkg/controller/resource.go [hunk 2]
```diff
@@ -137,10 +138,92 @@ func (c *Controller) ProcessFunc(ctx context.Context, resource runtime.Object, d
...
+		})
+
+		return errors.Wrapf(
+			c.applyAndCheck(ctx, r),
+			"apply %s (%s/%s)", gvk, namespace, name,
+		)
+	}
+}
+
+func needsHCNodePort(svc *corev1.Service) bool {
```
Comment (yxxhero) [inline pkg/controller/resource.go:209]: add unittest for this func?

File: pkg/controller/resource.go [hunk 3]
```diff
@@ -137,10 +138,92 @@ func (c *Controller) ProcessFunc(ctx context.Context, resource runtime.Object, d
...
+	}
+
+	if svc.Spec.ExternalTrafficPolicy != corev1.ServiceExternalTrafficPolicyTypeLocal {
+		return false
+	}
+
+	return true
+}
+
+func needsNodePort(svc *corev1.Service) bool {
```
Comment (yxxhero) [inline pkg/controller/resource.go:221]: add unittest for this func?

Comment (yxxhero) [reply pkg/controller/resource.go:221]: add comments will better.

**Review by yxxhero — APPROVED** [review]

**Pull request merged**